From 919080dca831e360e1b6ec1d4fbdab0d9cf0aa91 Mon Sep 17 00:00:00 2001
From: jarojarojaro-jaro <miki.jaroszek@gmail.com>
Date: Mon, 12 Oct 2026 19:31:21 +0000
Subject: [PATCH 18/19] CUDA/HIP: fix race in flash_attn_ext_f16_process_tile
 when nbatch_combine != DKQ/2 (cherry-pick of upstream #30103)

(cherry picked from commit 1cc47333a29b81487ea135f2ae473c21d2025e4a)
---
 ggml/src/ggml-cuda/fattn-mma-f16.cuh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ggml/src/ggml-cuda/fattn-mma-f16.cuh b/ggml/src/ggml-cuda/fattn-mma-f16.cuh
index 7f0f6abfd..296aa2b34 100644
--- a/ggml/src/ggml-cuda/fattn-mma-f16.cuh
+++ b/ggml/src/ggml-cuda/fattn-mma-f16.cuh
@@ -1772,7 +1772,7 @@ static __device__ __forceinline__ void flash_attn_ext_f16_process_tile(
                 }
             }
         }
-        if (np > 1) {
+        if (np > 1 || nbatch_combine != DKQ/2) {
             __syncthreads();
         }
     }
-- 
2.43.0

